cmd/outline-ss-server/expiring_map.go

```diff
@@ -0,0 +1,109 @@
+// Copyright 2024 The Outline Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"sync"
+	"sync/atomic"
+	"time"
+)
+
+// ExpiringMap is a thread-safe, generic map that automatically removes
+// key-value pairs after a specified duration of inactivity.
+// It employs reference counting to ensure safe concurrent access and prevent
+// premature deletion during cleanup.
+type ExpiringMap[K comparable, V any] struct {
```

Comment:
We are introducing significant complexity in order to cache and track the IP info, instead of leveraging its natural scope. It's not clear to me what the memory implications are for this too. You are adding new knobs that a server provider will need to care about.

I recommend we leverage the natural scope of the connection and NAT associations instead.

Here are a couple of ideas:

`AddOpenTCPConnection(clientAddr net.Addr)` could lookup the IP info and return a `TCPConnectionMetrics` object that holds the Ipinfo inside. The connection handler would hold the TCPConnectionMetrics. `AddClosedTCPConnection` would be called on ConnectionMetrics

UDP Would be similar, but on associations.

Instead of the handler creating the TCPConnectionMetrics, it could be passed to it in tcpHandler.Handle, alongside the connection.

Reply:
Not sure I fully agree it's a new knob they need to care about, but I gave this some more thought and I like your suggestion of making this a larger change to delegate the lookup. It changes the interfaces a bit. PTAL!